### DISCUSSION TITLE
risc0 `std` support and making Bonsai work with other VMs (not just EVM)

### DISCUSSION THREAD

**Discussion by IaroslavMazur** [risc0/risc0#1412] [Q&A] [answered]
Hello, `risc0` team & community!

I'm a member of the [Sablier](https://sablier.com/) team. We are currently evaluating various technologies for potential use in our future products, and `risc0` has caught our attention thanks to its promising features and capabilities.

For context: I've gone through the whole [RISC Zero @ ZK HACK III](https://www.youtube.com/playlist?list=PLcPzhUaCxlCgig7ofeARMPwQ8vbuD6hC5) workshop, as well as the [What does RISC V have to do with RISC Zero's zkVM](https://youtu.be/11DIflEwx50?feature=shared) presentation, the `risc0` [documentation](https://dev.risczero.com/api) (high-level) and several articles on the lower-level stuff (such as RISC-V and ZKPs).

Still, there are several technical uncertainties I'd like to clear out, in order to better understand how to properly use the `risc0` ZK technology.

### What's your current support and recommendations for the `std` Rust environment? 
In December of 2022 (when the RISC Zero @ ZK HACK III videos were released), the `std` support was deemed "experimental", but not discouraged (see this [time reference](https://youtu.be/Yg_BGqj_6lg?feature=shared&t=804)). One bug (related to the `String` data type) was mentioned specifically, alongside a workaround for it.

However, enough time has passed since then - and, maybe, you've already fixed the std-related problems. Therefore, my question: do you still consider the `risc0` `std` support experimental - or do we get the green light for safely using it?

### Can Bonsai only work with EVM?

Given that Layer 2s (with their ZK proof generation "on L2", followed by ZK verification on L1) are a common use-case for ZKPs, and, also, the following quote from the `Bonsai` documentation: [Bonsai enables boundless computation on any blockchain](https://dev.risczero.com/api/next/bonsai/), I assume that `Bonsai` isn't limited to only work with the Ethereum Virtual Machine, but can also be used with other VMs.

Could you confirm this, please?

Thank you in advance for your time and help!

**Comment by it09** [comment, marked as answer]
Hello Iaroslav! 
That is an excellent questions! 
**std support** 
We provide partial `std` support and are continue adding more feature support. While we are working on the docs to reflect current std support, I want to share this [file](https://github.com/rust-lang/rust/blob/master/library/std/src/sys/pal/zkvm/mod.rs) with you the way it's structured: the lines that say "unsupported" are things that we either don't support or we have different ways of supporting. We don't support threads, locking processes, pipes, networking, file systems, and time. It says here that we don't support io, but we have a different way of supporting them. 

**Can Bonsai only work with EVM?**
No. zkVM and Bonsai are for general purposes and are not limited to EVM. Our repo has more examples, here is a NEAR light client [example](https://github.com/risc0/demos/tree/main/near-zk-light-client). Also [bitcoin L2](https://docs.citrea.xyz/technical-specs/readme) is built on Risc0 zkVM.
Generally speaking, think of Risc0 zkVM as a superset to zkEVM. You can build zkEVM on zkVM, and we did it [here](https://github.com/risc0/zeth), and we did zeth in 4 weeks:))

**Reply by IaroslavMazur** [reply]
Hello, @it09, and thank you for your reply!

### std support

I've looked at the way system bindings are set for `risc0`'s `zkVM` and how the functions of the modules with limited support do nothing or panic with an error (via the `unsupported()` function).

With the above in mind, would it be correct to say that unless the `risc0` API malfunctions or throws that "unsupported" error, you shouldn't consider limiting your project (that is using the `risc0` technology) to just the `no-std` environment?

---

### Bonsai with any blockchain

Thank you for the confirmation and the relevant examples!

By the way, congrats on the speedy delivery of `zeth`!

**Reply by IaroslavMazur** [reply]
@it09, I've followed up with some extra (and more specific) questions via my Support Ticket in the `risc0` Discord

**Reply by it09** [reply]
some functionality that marked as unsupported in that file is actually supported via different means For example `io` is marked as unsupported, but we have a different support for it. What `std` functionality are you looking for?

**Reply by IaroslavMazur** [reply]
> What std functionality are you looking for?

Would you mind if we continue discussing the specifics on Discord?

Given the experimental nature of the `risc0`'s applicability to Sablier, it'd be best if we keep this private, until it materializes into something certain.
